Included Variables Don't Work in MQTT Block

```
# config.yaml
---

name: Common test information
description: Connection information for MQTT Client

variables:
  host: localhost
  port: 1883
```

```
# test_bug.tavern.yaml
---

test_name: Bug Demonstration

includes:
  - !include config.yaml

marks:
  - bugdemo

paho-mqtt:
  client:
    transport: websockets
    client_id: tavern-tester
  connect:
    host: "{host}"
    port: "{port:d}"
    timeout: 3

stages:

  - name: Step 1 - ping/pong
    mqtt_publish:
      topic: /device/123/ping
      payload: ping
    mqtt_response:
      topic: /device/123/pong
      payload: pong
      timeout: 5
```
The above test results in a schema error :-
```
------------------------------ Captured log call -------------------------------
core.py                    162 ERROR    validation.invalid
core.py                    163 ERROR     --- All found errors ---
core.py                    164 ERROR    [u"Value '{port}' is not of type 'int'. Path: '/paho-mqtt/connect/port'"]
files.py                   111 ERROR    Error validating {'paho-mqtt': {'client': {'client_id': 'tavern-tester', 'transport': 'websockets'}, 'connect': {'host': '{host}', 'port': '{port}', 'timeout': 3}}, 'marks': ['bugdemo'], 'stages': [{'mqtt_response': {'topic': '/device/123/pong', 'payload': 'pong', 'timeout': 5}, 'name': 'Step 1 - ping/pong', 'mqtt_publish': {'topic': '/device/123/ping', 'payload': 'ping'}}], 'includes': [{'variables': {'host': 'localhost', 'port': 1883}, 'name': 'Common test information', 'description': 'Connection information for MQTT Client'}], 'test_name': 'Bug Demonstration'}
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/tavern/schemas/files.py", line 109, in verify_generic
    verifier.validate()
  File "/usr/local/lib/python2.7/dist-packages/pykwalify/core.py", line 167, in validate
    error_msg=u'.\n - '.join(self.validation_errors)))
SchemaError: <SchemaError: error code 2: Schema validation failed:
 - Value '{port:d}' is not of type 'int'. Path: '/paho-mqtt/connect/port'.: Path: '/'>
```
Specifying the config file as tavern-global-cfg has the same result.

Sorry, Michael, I've returned too late.

It doesn't work, I'm afraid. I get the same error.

Furthermore, if I hard code the port value and just pass in the host, the test runs but fails to connect to the broker, which implies that the information isn't getting through.

```
test_spec = {'test_name': 'Bug Demonstration', 'includes': [{'name': 'Common test information', 'description': 'Connection informa...evice/123/ping', 'payload': 'ping'}, 'mqtt_response': {'topic': '/device/123/pong', 'payload': 'pong', 'timeout': 5}}]}
global_cfg = {'backends': {'http': 'requests', 'mqtt': 'paho-mqtt'}, 'strict': [], 'variables': {'host': '127.0.0.1', 'port': 1883...-terminal', '_': '/usr/local/bin/py.test', 'PYTEST_CURRENT_TEST': 'test_bug.tavern.yaml::Bug Demonstration (call)'}}>}}

    def run_test(in_file, test_spec, global_cfg):
        """Run a single tavern test
    
        Note that each tavern test can consist of multiple requests (log in,
        create, update, delete, etc).
    
        The global configuration is copied and used as an initial configuration for
        this test. Any values which are saved from any tests are saved into this
        test block and can be used for formatting in later stages in the test.
    
        Args:
            in_file (str): filename containing this test
            test_spec (dict): The specification for this test
            global_cfg (dict): Any global configuration for this test
    
        No Longer Raises:
            TavernException: If any of the tests failed
        """
    
        # pylint: disable=too-many-locals
    
        # Initialise test config for this test with the global configuration before
        # starting
        test_block_config = dict(global_cfg)
    
        if "variables" not in test_block_config:
            test_block_config["variables"] = {}
    
        tavern_box = Box({"env_vars": dict(os.environ)})
    
        if not test_spec:
            logger.warning("Empty test block in %s", in_file)
            return
    
        def stage_ids(s):
            return [i["id"] for i in s]
    
        available_stages = test_block_config.get("stages", [])
    
        if test_spec.get("includes"):
            # Need to do this separately here so there is no confusion between global and included stages
            for included in test_spec["includes"]:
                for stage in included.get("stages", {}):
                    if stage["id"] in stage_ids(available_stages):
                        msg = "Stage id '{}' defined in stage-included test which was already defined in global configuration - this will be an error in future!".format(
                            stage["id"]
                        )
                        logger.warning(msg)
                        warnings.warn(msg, FutureWarning)
    
            included_stages = []
    
            for included in test_spec["includes"]:
                if "variables" in included:
                    formatted_include = format_keys(
                        included["variables"], {"tavern": tavern_box}
                    )
                    test_block_config["variables"].update(formatted_include)
    
                for stage in included.get("stages", []):
                    if stage["id"] in stage_ids(included_stages):
                        raise exceptions.DuplicateStageDefinitionError(
                            "Stage with specified id already defined: {}".format(
                                stage["id"]
                            )
                        )
                    included_stages.append(stage)
        else:
            included_stages = []
    
        available_stages = {s["id"]: s for s in available_stages + included_stages}
    
        test_block_config["variables"]["tavern"] = tavern_box
    
        test_block_name = test_spec["test_name"]
    
        # Strict on body by default
        default_strictness = test_block_config["strict"]
    
        logger.info("Running test : %s", test_block_name)
    
        with ExitStack() as stack:
            test_spec["stages"] = _resolve_test_stages(test_spec, available_stages)
            sessions = get_extra_sessions(test_spec, test_block_config)
    
            for name, session in sessions.items():
                logger.debug("Entering context for %s", name)
>               stack.enter_context(session)

/usr/local/lib/python3.6/site-packages/tavern/core.py:136: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <contextlib2.ExitStack object at 0x7f7f4b4ac780>
cm = <tavern._plugins.mqtt.client.MQTTClient object at 0x7f7f4a223be0>

    def enter_context(self, cm):
        """Enters the supplied context manager
    
        If successful, also pushes its __exit__ method as a callback and
        returns the result of the __enter__ method.
        """
        # We look up the special methods on the type to match the with statement
        _cm_type = _get_type(cm)
        _exit = _cm_type.__exit__
>       result = _cm_type.__enter__(cm)

/usr/local/lib/python3.6/site-packages/contextlib2.py:380: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tavern._plugins.mqtt.client.MQTTClient object at 0x7f7f4a223be0>

    def __enter__(self):
        self._client.connect_async(**self._connect_args)
        self._client.loop_start()
    
        elapsed = 0
    
        while elapsed < self._connect_timeout:
            # pylint: disable=protected-access
            if self._client._state == paho.mqtt_cs_connected:
                logger.debug("Connected to broker at %s", self._connect_args["host"])
                return self
            else:
                logger.debug("Not connected after %s seconds - waiting", elapsed)
    
            # TODO
            # configurable?
            time.sleep(0.25)
            elapsed += 0.25
    
        self._disconnect()
        logger.error(
            "Could not connect to broker after %s seconds", self._connect_timeout
        )
>       raise exceptions.MQTTError
E       tavern.util.exceptions.MQTTError

/usr/local/lib/python3.6/site-packages/tavern/_plugins/mqtt/client.py:344: MQTTError
------------------------------ Captured log call -------------------------------
client.py                  342 ERROR    Could not connect to broker after 3 seconds
=================== 1 failed, 64 deselected in 3.85 seconds ====================
vagrant@vagrant-ubuntu-trusty-64:/media/sf_GIT_Repository/TDCPSimulator/Tests$ 

```

--- a/.gitignore
+++ b/.gitignore
@@ -111,5 +111,6 @@ pytype_output
 
 .vscode
 
+out/
 **/*.iml
 .idea

--- a/.pylintrc
+++ b/.pylintrc
@@ -1,5 +1,5 @@
 [MASTER]
-disable=missing-docstring,bad-continuation,fixme,invalid-name,line-too-long,too-few-public-methods,no-else-return,too-many-branches,locally-disabled,useless-object-inheritance
+disable=missing-docstring,bad-continuation,fixme,invalid-name,line-too-long,too-few-public-methods,no-else-return,too-many-branches,locally-disabled,useless-object-inheritance,no-else-raise
 ignore=tests
 
 [REPORTS]

--- a/example/mqtt/common.yaml
+++ b/example/mqtt/common.yaml
@@ -4,3 +4,4 @@ description: used for testing against local server
 
 variables:
   host: http://localhost:5002
+  mqtt_port: 5002

--- a/example/mqtt/common.yaml
+++ b/example/mqtt/common.yaml
@@ -4,4 +4,5 @@ description: used for testing against local server
 
 variables:
   host: http://localhost:5002
-  mqtt_port: 5002
+  mqtt_host: localhost
+  mqtt_port: 9001

--- a/example/mqtt/test_mqtt.tavern.yaml
+++ b/example/mqtt/test_mqtt.tavern.yaml
@@ -343,7 +343,7 @@ paho-mqtt:
     client_id: tavern-tester
   connect:
     host: "{host}"
-    port: "{mqtt_port:d}"
+    port: !int "{mqtt_port:d}"
     timeout: 3
 
 stages:

--- a/example/mqtt/test_mqtt.tavern.yaml
+++ b/example/mqtt/test_mqtt.tavern.yaml
@@ -342,17 +342,15 @@ paho-mqtt:
     transport: websockets
     client_id: tavern-tester
   connect:
-    host: "{host}"
+    host: "{mqtt_host}"
     port: !int "{mqtt_port:d}"
     timeout: 3
 
 stages:
-
-  - name: Step 1 - ping/pong
+  - name: step 1 - ping/pong
     mqtt_publish:
       topic: /device/123/ping
-      payload: ping
     mqtt_response:
       topic: /device/123/pong
-      payload: pong
       timeout: 5
+      qos: 1

--- a/example/mqtt/test_mqtt.tavern.yaml
+++ b/example/mqtt/test_mqtt.tavern.yaml
@@ -329,3 +329,30 @@ stages:
         message: !raw "Hello {world}"
       timeout: 5
       qos: 1
+
+---
+
+test_name: Test can use type token with mqtt port
+
+includes:
+  - !include common.yaml
+
+paho-mqtt:
+  client:
+    transport: websockets
+    client_id: tavern-tester
+  connect:
+    host: "{host}"
+    port: "{mqtt_port:d}"
+    timeout: 3
+
+stages:
+
+  - name: Step 1 - ping/pong
+    mqtt_publish:
+      topic: /device/123/ping
+      payload: ping
+    mqtt_response:
+      topic: /device/123/pong
+      payload: pong
+      timeout: 5

--- a/tavern/_plugins/mqtt/client.py
+++ b/tavern/_plugins/mqtt/client.py
@@ -319,6 +319,8 @@ class MQTTClient(object):
             )
 
     def __enter__(self):
+        logger.debug("Connecting to %s", self._connect_args)
+
         self._client.connect_async(**self._connect_args)
         self._client.loop_start()
 

--- a/tavern/_plugins/mqtt/schema.yaml
+++ b/tavern/_plugins/mqtt/schema.yaml
@@ -34,7 +34,8 @@ initialisation:
             type: str
           port:
             required: false
-            type: int
+            type: any
+            func: int_variable
           keepalive:
             required: false
             type: float

--- a/tavern/plugins.py
+++ b/tavern/plugins.py
@@ -8,6 +8,7 @@ import logging
 import stevedore
 from future.utils import raise_from
 
+from tavern.util.dict_util import format_keys
 from .util import exceptions
 
 logger = logging.getLogger(__name__)
@@ -154,7 +155,9 @@ def get_extra_sessions(test_spec, test_block_config):
             logger.debug(
                 "Initialising session for %s (%s)", p.name, p.plugin.session_type
             )
-            sessions[p.name] = p.plugin.session_type(**test_spec.get(p.name, {}))
+            session_spec = test_spec.get(p.name, {})
+            formatted = format_keys(session_spec, test_block_config.get("variables", {}))
+            sessions[p.name] = p.plugin.session_type(**formatted)
 
     return sessions
 

--- a/tavern/plugins.py
+++ b/tavern/plugins.py
@@ -156,7 +156,9 @@ def get_extra_sessions(test_spec, test_block_config):
                 "Initialising session for %s (%s)", p.name, p.plugin.session_type
             )
             session_spec = test_spec.get(p.name, {})
-            formatted = format_keys(session_spec, test_block_config.get("variables", {}))
+            formatted = format_keys(
+                session_spec, test_block_config.get("variables", {})
+            )
             sessions[p.name] = p.plugin.session_type(**formatted)
 
     return sessions

